packages/core/index.d.ts

```diff
@@ -17,51 +17,53 @@ interface PluginObject {
   requestEnd?: PluginHookPromise
 }
 
-interface Request<TEvent = any, TResult = any, TErr = Error> {
+interface Request<TEvent = any, TResult = any, TErr = Error, TContext extends LambdaContext = LambdaContext> {
   event: TEvent
-  context: LambdaContext
+  context: TContext
   response: TResult | null
   error: TErr | null
   internal: {
     [key: string]: any
   }
 }
 
-declare type MiddlewareFn<TEvent = any, TResult = any, TErr = Error> = (request: Request<TEvent, TResult, TErr>) => any
+declare type MiddlewareFn<TEvent = any, TResult = any, TErr = Error, TContext extends LambdaContext = any> = (request: Request<TEvent, TResult, TErr, TContext>) => any
 
-interface MiddlewareObj<TEvent = any, TResult = any, TErr = Error> {
-  before?: MiddlewareFn<TEvent, TResult, TErr>
-  after?: MiddlewareFn<TEvent, TResult, TErr>
+interface MiddlewareObj<TEvent = any, TResult = any, TErr = Error, TContext extends LambdaContext = any> {
+  before?: MiddlewareFn<TEvent, TResult, TErr, TContext>
+  after?: MiddlewareFn<TEvent, TResult, TErr, TContext>
   onError?: MiddlewareFn<TEvent, TResult, TErr>
 }
 
-interface MiddyfiedHandler<TEvent = any, TResult = any, TErr = Error> {
-  use: UseFn<TEvent, TResult, TErr>
-  applyMiddleware: AttachMiddlewareObj<TEvent, TResult, TErr>
-  before: AttachMiddlewareFn<TEvent, TResult, TErr>
-  after: AttachMiddlewareFn<TEvent, TResult, TErr>
-  onError: AttachMiddlewareFn<TEvent, TResult, TErr>
+interface MiddyfiedHandler<TEvent = any, TResult = any, TErr = Error, TContext extends LambdaContext = any> {
```

Comment:
From discussion in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52644 I would suggest something like (untested):

```suggestion
// Should accept either LambdaHandlers or MiddyfiedHandlers.
interface MiddyInputHandler<TEvent, TResult, TContext extends LambdaContext> extends LambdaHandler<TEvent, TResult> {
    (event: TEvent, context: TContext) => Promise<TResult> | void;
}

interface MiddyfiedHandler<TEvent = any, TResult = any, TErr = Error, TContext extends LambdaContext = any> extends MiddyInputHandler<TEvent, TResult, TContext> {
```

Then using `MiddyInputHandler<>` instead of `LambdaHandler<>` below.

Reply:
Unfortunately, the `extends LambdaHandler<>` [seems to break the type?](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIHtwQB5mQbwFgAoZM5BTSXANTgBsBXCALmREYFsAjaAbhIBfEiTABPAA4oAEnBAATetAA8AFQCiANwjgANMlUAlCAGdG9MAD5kAXmQAKCNvBsNzsPspZcbDN7AAlLbWAApQ6JzAJhBqxmYWlgLEJKCQsIgoAMoREFo6eESk5E75dEysyCZgUKAA5knCxKnQ8EjI2ZwQceYFJOTIUKY9Zcxs3OjoSnINopJZObIKSlC2yIuKKh257vpb3QlJKVjpbQCywPLyYgCSIBKMYOvLanl6BvseBn7UeDiQCiY0FQ-tY-jp5ICnio3Pl9EYhgkCH1yI53K5Xp8vD9XN8-gE2GEIlEYvD4lZkAAfZCadAXGbJJrHVooc6XJQAdzgg1xuGQYIBQP8SKKZEibIgnMGIwqJjEPEm9PEUnaOVZG0lEChKzsapudweWuUWwxuxyH30ao5XIgPKsh2IXiqyAAFnINlA2FtLRLrVrVnBZSAEA4Sm8sXjgsL+kpfu5pWwqjUQLVVqGwAA6NPSpLRiB4cO0BijdhcXjaijA3DpgtgbPIsgx5Bi9XW+OVOXjeirGvp5tWqVFiA58iDMCMKAgKP9EcI2uDtjVZi6evIQQNJJAA)

Reply:
Thanks for sharing that example, I definitely agree with your suggestion.

```
interface Context {
    contextValue: number;
}

type Handler<TEvent, TResult> = (event: TEvent, context: Context) => Promise<TResult>;

interface SomeEvent {
    eventValue: string;
}
interface SomeResult {
    resultValue: boolean;
}
type SomeHandler = Handler<SomeEvent, SomeResult>;

interface MiddyInputHandler<TEvent, TResult, TContext extends Context> extends Handler<TEvent, TResult> {
    (event: TEvent, context: TContext): Promise<TResult> | void;
}

interface MiddlewareContext extends Context {
    middlewareValue: symbol;
}
type SomeMiddlewareHandler = MiddyInputHandler<SomeEvent, SomeResult, MiddlewareContext>;

const handler: SomeMiddlewareHandler = async (event, context) => {
    let eventValue: string = event.eventValue;
    let contextValue: number = context.contextValue;
    let middlewareValue: symbol = context.middlewareValue;
    return {
        resultValue: true,
    };
};
```

`const handler: SomeMiddlewareHandler = async (event, context) => { ... }` fails due to the parameters `event` and `context` having implicit type `any`

I think the simplest fix for this is adding by declaring a type for MiddyInputHandler
```
type MiddyInputHandlerT<TEvent, TResult, TContext extends Context> = (event: TEvent, context: TContext) => Promise<TResult> | void; 
```

Then my modifying the type of `SomeMiddlewareHandler` we can resolve the implicit type errors in `handler`
```
type SomeMiddlewareHandler<TContext extends MiddlewareContext = any> = MiddyInputHandlerT<SomeEvent, SomeResult, TContext>;
```

Reply:
[result](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIHtwQB5mQbwFgAoZM5BTSXANTgBsBXCALmREYFsAjaAbhIBfEiTABPAA4oAEnBAATetAA8AFQCiANwjgANMlUAlCAGdG9MAD5kAXmQAKCNvBsNzsPspZcbDN7AAlLbWAApQ6JzAJhBqxmYWlgLEJKCQsIgoAMoREFo6eESk5E75dEysyCZgUKAA5knCxKnQ8EjI2ZwQceYFJOTIUKY9Zcxs3OjoSnINopJZObIKSlC2yIuKKh257vpb3QlJKVjpbQCywPLyYgCSIBKMYOvLanl6BvseBn7UeDiQCiY0FQ-tY-jp5ICnio3Pl9EYhgkCH1yI53K5Xp8vD9XN8-gE2GEIlEYvD4lZkAAfZCadAXGbJYjiKTIc6XG53B5QqCqF47d4Iz6qXG4ZBggFA-zWOyo-LovlYv444G4II2ULhSLRWIC6xUml05AiJrHVooVkbADucEGwt+uHBgNtSKKZEilyUVsGIwqJjEPEm9KZ8065o91ogXLUTrFEJZF0t4addjkYilcbZt3ujzkG25yi2GN2OQ+cNtiSNXiqyAAFjnlmwtqGIJ6I3XoKs4L6QAgHCU3gqVcFnf0lL93N62FUaiBaqs+2AAHTz71JEcQPADsAT9hcXgrOybheblfIsij5BuhNehijSp+8b0VaHy9h6-lVfkQZgRhQEDD-qfgK27VMwuinsgggNIcDJAA)